analyses/doublet-detection/exploratory-notebooks/03_compare-benchmark-results.Rmd

```diff
@@ -0,0 +1,363 @@
+        dplyr::rowwise() |>
+        dplyr::mutate(consensus_call = dplyr::if_else(
+          all(
+            c(scdbl_prediction, scrublet_prediction, cxds_prediction) == "doublet"
+          ),
+          "doublet", 
+          "singlet"
+        )) |>
+        dplyr::mutate(
+          call_type = dplyr::case_when(
+            consensus_call == "doublet" && ground_truth == "doublet" ~ "tp",
+            consensus_call == "singlet" && ground_truth == "singlet" ~ "tn",
+            consensus_call == "doublet" && ground_truth == "singlet" ~ "fp",
+            consensus_call == "singlet" && ground_truth == "doublet" ~ "fn"
+          ), 
+          # set associated plotting colors
+          call_type_color = dplyr::case_when(
+            call_type == "tp" ~ unname(confusion_colors["tp"]),
+            call_type == "tn" ~ unname(confusion_colors["tn"]),
+            call_type == "fp" ~ unname(confusion_colors["fp"]),
+            call_type == "fn" ~ unname(confusion_colors["fn"])
+          )
+          )
+      
+      return(dataset_df)
+    }
+  )
+names(doublet_df_list) <- dataset_names
+```
+
+
+
+
+## Upset plots
+
+This section contains upset plots that show overlap across doublet calls from each method, displayed for each dataset.
+
+```{r}
+pull_barcodes <- function(df, pred_var) {
+  # Helper function to pull out barcodes for doublet calls
+  df$barcodes[df[[pred_var]] == "doublet"]
+}
+
+upset_list <- doublet_df_list |>
+  purrr::iwalk(
+    \(df, dataset) {
+      
+      doublet_barcodes <- list(
+        "scDblFinder" = pull_barcodes(df, "scdbl_prediction"),
+        "scrublet"    = pull_barcodes(df, "scrublet_prediction"),
+        "cxds"        = pull_barcodes(df, "cxds_prediction")
+      )
+      
+      UpSetR::upset(fromList(doublet_barcodes), order.by = "freq") |> print()
+      grid::grid.text( # plot title
+        dataset,
+        x = 0.65, 
+        y = 0.95, 
+        gp = grid::gpar(fontsize=16)
+      ) 
+
+    }
+  )
+
+```
+
+
+
+## Evaluating consensus performance
+
+This section visualizes and evaluates the consensus doublet calls across each dataset.
+
+
+
+### PCA
+
+This section plots the PCA for each dataset, with three color schemes from left to right:
+
+1. Ground truth doublets are shown in black and singlets in blue
+2. Consensus doublets are shown in black and all remaining droplets in blue
+3. Points are colored based on comparing the consensus call to the ground truth as one of:
+  - true positive (`tp`), true negative (`tn`), false positive (`fp`), false negative (`fn`)
+
+
+```{r, fig.width = 12, fig.height = 6}
+# Make a legend for the confusion-colored PCA
+legend_plot <- data.frame(
+  x = factor(names(confusion_colors), levels = names(confusion_colors)), y = 1:4
+) |>
+ ggplot(aes(x = x, y = y, color = x)) + 
+  geom_point(size = 3) + 
+  scale_color_manual(name = "Metric", values = confusion_colors) 
+confusion_legend <- ggpubr::get_legend(legend_plot) |> ggpubr::as_ggplot()
```

Comment:
Why do you need to do this separately and not within the function for creating the confusion plot? Also why do you need to do this at all... the legend should automatically match the colors that you specify which you can do with `scale_color_manual` and a named list of values.

Reply:
Hm.... I did this because not all the plots had all of the four colors, but using a _named list_ never crossed my mind........................... That is of course better!